internal/types/compare.go

```diff
@@ -306,3 +295,83 @@ func compareArrays(filterArr, docArr *Array) CompareResult {
 
 	return Equal
 }
+
+// compareDocuments compares documents recursively by
+// comparing them in the order of types, field names and field values.
+func compareDocuments(a, b *Document) CompareResult {
+	if a.Len() == 0 && b.Len() == 0 {
+		return Equal
+	}
+
+	if a.Len() == 0 && b.Len() > 0 {
+		return Less
+	}
+
+	if b.Len() == 0 {
+		return Greater
+	}
+
+	aKeys := a.Keys()
+	bKeys := b.Keys()
+	bValues := b.Values()
+	aValues := a.Values()
+
+	for i, aKey := range aKeys {
+		if b.Len() == i {
+			return Greater
+		}
+
+		// compare type
+		if result := compareTypeOrder(aValues[i], bValues[i]); result != Equal {
+			return result
+		}
+
+		// compare keys
+		if result := compareScalars(aKey, bKeys[i]); result != Equal {
+			return result
+		}
+
+		// compare values
+		if result := Compare(aValues[i], bValues[i]); result != Equal {
+			return result
+		}
+	}
+
+	if a.Len() < b.Len() {
+		return Less
+	}
+
+	return Equal
+}
+
+// compareArrayToNonArray compares array to a non-array value.
+func compareArrayToNonArray(as *Array, b any) CompareResult {
```

Comment:
What happens if b is an array too? If the function doesn't support that case, should we panic?

Reply:
Maybe I can call `compareArrays(as,b)`. At the moment it might panic :scream:

Reply:
I mean, it's fine if we panic in such a case, if we expect the caller to be responsible for `b` not being the array, let's make such a panic explicit

Reply:
Actually since the code can handle comparing two arrays, I think it makes sense for this function to call `compareArrays` to compare them array. No need to panic here.